runtime/near-vm-logic/src/logic.rs

```diff
@@ -1541,7 +1594,27 @@ impl<'a> VMLogic<'a> {
 
         self.deduct_balance(amount)?;
 
-        self.ext.append_action_function_call(receipt_idx, method_name, arguments, amount, gas)?;
+        #[cfg(feature = "protocol_feature_function_call_weight")]
+        self.ext.append_action_function_call_weight(
+            receipt_idx,
+            method_name,
+            arguments,
+            amount,
+            gas,
+            gas_weight,
+        )?;
+
+        #[cfg(not(feature = "protocol_feature_function_call_weight"))]
+        {
+            let _ = gas_weight;
+            self.ext.append_action_function_call(
```

Comment:
I just did this to avoid code duplication from `promise_batch_action_function_call`. Are you suggesting to panic if the weight is something other than 0 without the protocol feature?

Reply:
For deduplication I would probably add an inherent method which takes a callback/closure to call at the end.

So e.g.

```rust
    fn promise_batch_action_function_call_generic<F: FnOnce(...) -> Result<()>>(
        &mut self, 
        promise_idx: u64,
        method_name_len: u64,
        method_name_ptr: u64,
        arguments_len: u64,
        arguments_ptr: u64,
        amount_ptr: u64,
        gas: Gas,
        callback: F,
    ) -> Result<()> {
        /* the deduplicated logic... */
        callback()
    }

    pub fn promise_batch_action_function_call(
        &mut self,
        promise_idx: u64,
        method_name_len: u64,
        method_name_ptr: u64,
        arguments_len: u64,
        arguments_ptr: u64,
        amount_ptr: u64,
        gas: Gas,
    ) -> Result<()> {
        self.promise_batch_action_function_call_generic(
            promise_idx,
            method_name_len,
            method_name_ptr,
            arguments_len,
            arguments_ptr,
            amount_ptr,
            gas,
            |...| {
                self.ext.append_action_function_call(...)
            }
        )
    }

    #[cfg(feature = "protocol_feature_function_call_weight")]
    pub fn promise_batch_action_function_call_weight(...) {
        self.promise_batch_action_function_call_generic(..., |...| {
            self.ext.append_action_function_call_weight(...)
        })
    }
```

But to answer your question, the fact that a call to `promise_batch_action_function_call_weight` can be called with or without this protocol feature enabled seems pretty sketchy to me, as well as the fact that it transparently can choose to behave exactly like `promise_batch_action_function_call`.